akka-remote/src/main/scala/akka/remote/artery/SSLEngineProvider.scala

```diff
@@ -0,0 +1,222 @@
+        new SecureRandom
+
+      case unknown ⇒
+        log.warning(LogMarker.Security, "Unknown SSLRandomNumberGenerator [{}] falling back to SecureRandom", unknown)
+        new SecureRandom
+    }
+    rng.nextInt() // prevent stall on first access
+    rng
+  }
+
+  private def createSSLEngine(role: TLSRole): SSLEngine = {
+    createSSLEngine(getOrCreateContext(), sslConfig = None, role)
+  }
+
+  override def createServerSSLEngine(): SSLEngine =
+    createSSLEngine(akka.stream.Server)
+
+  override def createClientSSLEngine(): SSLEngine =
+    createSSLEngine(akka.stream.Client)
+
+  private def createSSLEngine(
+    sslContext: SSLContext,
+    sslConfig:  Option[AkkaSSLConfig],
+    role:       TLSRole,
+    closing:    TLSClosing            = IgnoreComplete,
+    hostInfo:   Option[(String, Int)] = None): SSLEngine = {
+
+    /* FIXME this is what NettySSLSupport does
+    val sslEngine = settings.getOrCreateContext(log).createSSLEngine // TODO: pass host information to enable host verification
+    sslEngine.setUseClientMode(isClient)
+    sslEngine.setEnabledCipherSuites(settings.SSLEnabledAlgorithms.toArray)
+    sslEngine.setEnabledProtocols(Array(settings.SSLProtocol))
+    */
+
+    // below is based on how stream.TLS creates SSLEngine from SSLContext
+
+    val firstSession = NegotiateNewSession.withCipherSuites(SSLEnabledAlgorithms.toSeq: _*)
+
+    def theSslConfig(system: ActorSystem): AkkaSSLConfig =
+      sslConfig.getOrElse(AkkaSSLConfig(system))
+
+    val engine = hostInfo match {
+      case Some((hostname, port)) ⇒ sslContext.createSSLEngine(hostname, port)
+      case None                   ⇒ sslContext.createSSLEngine()
+    }
+    val config = theSslConfig(system)
+    config.sslEngineConfigurator.configure(engine, sslContext)
+
+    engine.setUseClientMode(role == akka.stream.Client)
+
+    // FIXME took this from NettySSLSupport, still valid?
+    if ((role != akka.stream.Client) && SSLRequireMutualAuthentication) engine.setNeedClientAuth(true)
+
+    val finalSessionParameters =
+      if (firstSession.sslParameters.isDefined && hostInfo.isDefined && !config.config.loose.disableSNI) {
+        val newParams = TlsUtils.cloneParameters(firstSession.sslParameters.get)
+        // In Java 7, SNI was automatically enabled by enabling "jsse.enableSNIExtension" and using
+        // `createSSLEngine(hostname, port)`.
+        // In Java 8, SNI is only enabled if the server names are added to the parameters.
+        // See https://github.com/akka/akka/issues/19287.
+        newParams.setServerNames(Collections.singletonList(new SNIHostName(hostInfo.get._1)))
+        firstSession.copy(sslParameters = Some(newParams))
+      } else
+        firstSession
+
+    TlsUtils.applySessionParameters(engine, finalSessionParameters)
+
+    /* FIXME what about this, parameter to TlsModule?
+    def verifySession: (ActorSystem, SSLSession) ⇒ Try[Unit] =
```

Comment:
That's more backwards compat code to support `AkkaSSLConfig`. Not needed here.

Reply:
Though, `verifySession` could be exposed later on in some form to allow users to specify custom verifiers to use after a TLS session was established, like looking at concrete ciphers, certificates, etc.